File: ch_item_master/ch_item_master/page/ch_logistics_hub/ch_logistics_hub.js
/* CH Logistics Hub — Online/Bot Claim Pickup Queue
   Shows only Online/Bot and Phone channel claims in pickup stages.
   Walk-in / Store claims are handled at the store and never appear here. */

frappe.pages["ch-logistics-hub"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Logistics Hub"),
		single_column: true,
	});
	wrapper._hub = new LogisticsHub(page);
};

frappe.pages["ch-logistics-hub"].refresh = function (wrapper) {
	wrapper._hub && wrapper._hub.refresh();
};

const STATUS_COLORS = {
	"Pickup Requested":  "orange",
	"Pickup Scheduled":  "blue",
	"Picked Up":         "purple",
	"Device Received":   "green",
};

class LogisticsHub {
	constructor(page) {
		this.page = page;
		this._setup_filters();
		this._setup_container();
		this.refresh();
	}

	_setup_filters() {
		this.status_field = this.page.add_field({
			fieldname: "status", label: __("Status"), fieldtype: "MultiSelectList",
			options: ["Pickup Requested","Pickup Scheduled","Picked Up","Device Received"],
			default: [],
			change: () => this.refresh(),
		});
		this.company_field = this.page.add_field({
			fieldname: "company", label: __("Company"), fieldtype: "Link", options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			change: () => this.refresh(),
		});
		this.page.add_button(__("Refresh"), () => this.refresh(), { icon: "refresh" });
	}

	_setup_container() {
		this.$root = $(`<div class="lh-root" style="padding:16px 20px;"></div>`)
			.appendTo(this.page.body);
	}

	refresh() {
		const filters = {
			company: this.company_field?.get_value() || "",
			status: this.status_field?.get_value() || [],
		};
		this.$root.html(`<div style="padding:40px;text-align:center;color:var(--text-muted)">
			<i class="fa fa-spinner fa-spin"></i> ${__("Loading...")}
		</div>`);
		frappe.xcall(
			"ch_item_master.ch_item_master.page.ch_logistics_hub.ch_logistics_hub.get_logistics_queue",
			{ filters }
		).then(rows => this._render(rows))
		 .catch(() => {
			this.$root.html(`<div style="padding:40px;text-align:center;color:var(--error)">
				${__("Failed to load queue. Please retry.")}
			</div>`);
		});
	}

	_render(rows) {
		if (!rows.length) {
			this.$root.html(`<div style="padding:60px;text-align:center;color:var(--text-muted);font-size:13px;">
				<i class="fa fa-inbox" style="font-size:32px;display:block;margin-bottom:10px;opacity:.4;"></i>
				${__("No claims in the logistics queue.")}
			</div>`);
			return;
		}

		const cards = rows.map(r => this._card(r)).join("");
		this.$root.html(`
			<div style="margin-bottom:12px;font-size:12px;color:var(--text-muted);">
				${rows.length} ${__("claim(s) in queue")}
			</div>
			<div class="lh-grid" style="display:grid;grid-template-columns:repeat(auto-fill,minmax(340px,1fr));gap:14px;">
				${cards}
			</div>
		`);
		this.$root.find(".lh-action-btn").on("click", (e) => {
			const btn = $(e.currentTarget);
			const action = btn.data("action");
			const claim  = btn.data("claim");
			this[`_do_${action}`](claim, btn);
		});
	}

	_card(r) {
		const color = STATUS_COLORS[r.claim_status] || "grey";
		const slot  = r.pickup_slot ? frappe.datetime.str_to_user(r.pickup_slot) : "—";
		const actions = this._actions_html(r);
		return `
		<div class="lh-card" style="background:var(--card-bg);border:1px solid var(--border-color);
		     border-radius:8px;padding:14px;position:relative;">
			<div style="display:flex;align-items:center;gap:8px;margin-bottom:10px;">
				<span class="indicator-pill ${color}" style="flex-shrink:0;"></span>
				<a href="/app/ch-warranty-claim/${r.name}" target="_blank"
				   style="font-weight:700;font-size:13px;">${r.name}</a>
				<span style="margin-left:auto;font-size:10px;color:var(--text-muted);
				      background:var(--subtle-fg);padding:2px 8px;border-radius:4px;">
					${__(r.claim_channel)}
				</span>
			</div>
			<div style="font-size:12px;line-height:1.8;color:var(--text-color);">
				<b>${__("Customer")}:</b> ${frappe.utils.escape_html(r.customer_name || r.customer)} &nbsp;
				<span style="color:var(--text-muted);">${r.customer_phone || ""}</span><br>
				<b>${__("Device")}:</b> ${frappe.utils.escape_html(r.brand || "")} ${frappe.utils.escape_html(r.item_code || "")}
				 &nbsp; <code style="font-size:11px;">${r.serial_no || ""}</code><br>
				<b>${__("Slot")}:</b> ${slot}<br>
				<b>${__("Partner")}:</b> ${frappe.utils.escape_html(r.pickup_partner || "—")}<br>
				<b>${__("Tracking")}:</b> ${frappe.utils.escape_html(r.pickup_tracking_no || "—")}<br>
				<b>${__("Destination")}:</b> ${__(r.repair_destination || "Not Set")}
			</div>
			<div style="margin-top:6px;">
				${r.parcel_photo_1 ? `<a href="${r.parcel_photo_1}" target="_blank" style="font-size:11px;margin-right:6px;">📦 ${__("Parcel Photo 1")}</a>` : ""}
				${r.parcel_photo_2 ? `<a href="${r.parcel_photo_2}" target="_blank" style="font-size:11px;">📦 ${__("Parcel Photo 2")}</a>` : ""}
			</div>
			<div style="margin-top:12px;display:flex;flex-wrap:wrap;gap:6px;">
				${actions}
			</div>
		</div>`;
	}

	_actions_html(r) {
		const btn = (label, action, style="") =>
			`<button class="btn btn-xs btn-default lh-action-btn"
			         data-action="${action}" data-claim="${r.name}"
			         style="${style}">${__(label)}</button>`;
		const s = r.claim_status;
		let html = "";
		if (s === "Pickup Requested")  html += btn("Schedule Pickup", "schedule_pickup", "border-color:#f59e0b;color:#b45309;");
		if (s === "Pickup Scheduled")  html += btn("Mark Picked Up", "picked_up", "border-color:#8b5cf6;color:#6d28d9;");
		if (s === "Picked Up")         html += btn("Mark Device Received", "device_received", "border-color:#10b981;color:#065f46;");
		html += btn("View Claim", "view_claim", "");
		return html;
	}

	_do_schedule_pickup(claim, btn) {
		const d = new frappe.ui.Dialog({
			title: __("Schedule Pickup — {0}", [claim]),
			fields: [
				{ fieldname: "pickup_slot", fieldtype: "Datetime", label: __("Pickup Slot"), reqd: 1 },
				{ fieldname: "pickup_partner", fieldtype: "Data", label: __("Logistics Partner") },
				{ fieldname: "pickup_tracking_no", fieldtype: "Data", label: __("Tracking Number") },
			],
			primary_action_label: __("Schedule"),
			primary_action: (v) => {
				frappe.xcall(
					"ch_item_master.ch_item_master.page.ch_logistics_hub.ch_logistics_hub.schedule_pickup",
					{ claim_name: claim, ...v }
				).then(() => {
					d.hide();
					frappe.show_alert({ message: __("Pickup scheduled"), indicator: "green" });
					this.refresh();
				});
			},
		});
		d.show();
	}

	_do_picked_up(claim) {
		const d = new frappe.ui.Dialog({
			title: __("Mark Picked Up — {0}", [claim]),
			fields: [
				{ fieldname: "html_warn", fieldtype: "HTML",
				  options: `<div class="alert alert-warning" style="font-size:12px;">
				    ${__("Upload parcel photos BEFORE clicking Confirm for Online/Bot claims.")}
				  </div>` },
				{ fieldname: "delivery_otp", fieldtype: "Data", label: __("Delivery OTP") },
				{ fieldname: "remarks", fieldtype: "Small Text", label: __("Remarks") },
			],
			primary_action_label: __("Confirm Pickup"),
			primary_action: (v) => {
				frappe.xcall(
					"ch_item_master.ch_item_master.page.ch_logistics_hub.ch_logistics_hub.mark_picked_up",
					{ claim_name: claim, delivery_otp: v.delivery_otp, remarks: v.remarks }
				).then(() => {
					d.hide();
					frappe.show_alert({ message: __("Marked as picked up"), indicator: "green" });
					this.refresh();
				}).catch(err => {
					frappe.msgprint({ title: __("Error"), message: err.message || err, indicator: "red" });
				});
			},
		});
		d.show();
	}

	_do_device_received(claim) {
		const d = new frappe.ui.Dialog({
			title: __("Mark Device Received — {0}", [claim]),
			fields: [
				{ fieldname: "condition_on_receipt", fieldtype: "Select", label: __("Device Condition"),
				  options: "\nGood\nMinor Damage\nMajor Damage\nWrong Device\nEmpty Parcel\nIMEI Mismatch\nMissing Accessories",
				  reqd: 1 },
				{ fieldname: "accessories_received", fieldtype: "Data", label: __("Accessories Received") },
				{ fieldname: "imei_verified", fieldtype: "Check", label: __("IMEI Verified") },
				{ fieldname: "receiving_remarks", fieldtype: "Small Text", label: __("Remarks") },
			],
			primary_action_label: __("Confirm Receipt"),
			primary_action: (v) => {
				frappe.xcall(
					"ch_item_master.ch_item_master.page.ch_logistics_hub.ch_logistics_hub.mark_device_received",
					{ claim_name: claim, ...v }
				).then(() => {
					d.hide();
					frappe.show_alert({ message: __("Device received"), indicator: "green" });
					this.refresh();
				});
			},
		});
		d.show();
	}

	_do_view_claim(claim) {
		frappe.set_route("Form", "CH Warranty Claim", claim);
	}
}

File: ch_item_master/ch_item_master/page/ch_logistics_hub/ch_logistics_hub.py
import frappe
from frappe import _

from ch_item_master.config import get_int_setting, require_role_setting


_LOGISTICS_ROLES = ("CH Warranty Manager", "Service Manager", "Sales Manager")
_QUEUE_STATUSES = frozenset(
    {"Pickup Requested", "Pickup Scheduled", "Picked Up", "Device Received"}
)


def _require_logistics_access(action):
    require_role_setting("warranty_claim_logistics_roles", _LOGISTICS_ROLES, action=action)
    frappe.has_permission("CH Warranty Claim", "read", throw=True)


@frappe.whitelist()
def get_logistics_queue(filters=None):
    """Return online/phone claims in pickup-stage for logistics team.

    Only claims with claim_channel IN ('Online/Bot', 'Phone') and in a
    pickup-stage status are shown here. Store/walk-in claims are NOT shown.
    """
    _require_logistics_access(_("view the warranty logistics queue"))
    filters = filters or {}
    if isinstance(filters, str):
        try:
            filters = frappe.parse_json(filters) or {}
        except Exception:
            frappe.throw(_("Invalid logistics queue filters."))
    if not isinstance(filters, dict):
        frappe.throw(_("Invalid logistics queue filters."))

    status_filter = filters.get("status") or sorted(_QUEUE_STATUSES)
    if isinstance(status_filter, str):
        status_filter = [status_filter]
    status_filter = [status for status in status_filter if status in _QUEUE_STATUSES]
    if not status_filter:
        return []

    try:
        from ch_erp15.ch_erp15.scope import intersect_filters
    except (ImportError, ModuleNotFoundError):
        frappe.throw(_("Location scope validation is unavailable."), frappe.PermissionError)

    effective = intersect_filters(
        company=filters.get("company"),
        city=filters.get("city"),
        zone=filters.get("zone"),
        store=filters.get("store"),
    )

    channel_filter = ["Online/Bot", "Phone"]

    cond_parts = [
        "c.docstatus = 1",
        "c.claim_channel IN %(channels)s",
        "c.claim_status IN %(statuses)s",
    ]
    params = {
        "channels": tuple(channel_filter),
        "statuses": tuple(status_filter),
    }
    if effective.get("company"):
        cond_parts.append("c.company = %(company)s")
        params["company"] = effective["company"]
    allowed_stores = effective.get("allowed_stores")
    if allowed_stores is not None:
        if not allowed_stores:
            return []
        cond_parts.append("c.reported_at_store IN %(stores)s")
        params["stores"] = tuple(allowed_stores)

    limit = min(get_int_setting("warranty_logistics_queue_limit", 200, minimum=1), 1000)
    params["limit"] = limit

    rows = frappe.db.sql(
        """
        SELECT
            c.name, c.claim_id, c.claim_status, c.claim_channel,
            c.customer, c.customer_name, c.customer_phone, c.customer_email,
            c.serial_no, c.item_code, c.brand,
            c.pickup_address, c.pickup_slot, c.pickup_partner,
            c.pickup_tracking_no, c.picked_up_at, c.device_received_at,
            c.parcel_photo_1, c.parcel_photo_2,
            c.device_received_photo_1, c.device_received_photo_2,
            c.logistics_status, c.repair_destination,
            c.creation, c.modified
        FROM `tabCH Warranty Claim` c
        WHERE """
        + " AND ".join(cond_parts)
        + " ORDER BY c.creation ASC LIMIT %(limit)s",
        params,
        as_dict=True,
    )
    return rows


@frappe.whitelist(methods=["POST"])
def schedule_pickup(claim_name, pickup_slot, pickup_partner=None, pickup_tracking_no=None):
    """Schedule pickup for an online/bot claim."""
    doc = frappe.get_doc("CH Warranty Claim", claim_name)
    return doc.schedule_pickup(
        pickup_slot=pickup_slot,
        pickup_partner=pickup_partner,
        pickup_tracking_no=pickup_tracking_no,
    )


@frappe.whitelist(methods=["POST"])
def mark_picked_up(claim_name, delivery_otp=None, remarks=None):
    """Mark device as picked up — enforces photo gate for online claims."""
    doc = frappe.get_doc("CH Warranty Claim", claim_name)
    return doc.mark_picked_up(delivery_otp=delivery_otp, remarks=remarks)


@frappe.whitelist(methods=["POST"])
def mark_device_received(
    claim_name,
    condition_on_receipt,
    accessories_received=None,
    imei_verified=None,
    receiving_remarks=None,
):
    """Mark device as received at hub after pickup."""
    doc = frappe.get_doc("CH Warranty Claim", claim_name)
    return doc.mark_device_received(
        condition_on_receipt=condition_on_receipt,
        accessories_received=accessories_received,
        imei_verified=imei_verified,
        receiving_remarks=receiving_remarks,
    )
